### PR TITLE
Bug fix and many improvements

### PR DESCRIPTION
**Pull request by gfr10598**
This fixes the bug in the earlier "improved" sidestream.
It also:
  Creates a Web100StatsWriter class.
  Adds last 6 bits of local IP to connection counts.
  Refactors for readability
  Adds some unit tests (not enough - I'm writing more while you review this to get the coverage up higher).
  Reformats the production code.  (I'll reformat the test code in a followup CL, so that review is easier.)

---
This change is [Reviewable](https://reviewable.io/reviews/npad/sidestream/26)

### CODE REVIEW

File: exitstats.py [hunk 1]
```diff
@@ -17,214 +19,258 @@
...
-    logf.write(" "+str(snap[v]))
-  logf.write("\n")
-  logf.flush()
+                                ['type', 'lsb'])
+exception_count = prom.Counter('sidestream_exception_count',
+                               'Count of exceptions.',
+                               ['type'])
+
+class Web100StatsWriter:
+    ''' TODO - add documentation.
```
Comment (pboothe) [inline exitstats.py]: The eternal TODO

File: exitstats.py [hunk 2]
```diff
@@ -17,214 +19,258 @@
...
+            try:
+                if c.read('State') == 1:
+                    newclosed.append(c.cid)
+                    if not c.cid in closed:
+                        stats_writer.logConnection(c)
+            except Exception, e:
+                # We should handle all exceptions deeper in the call stack.
+                # We instrument this so that we can detect exceptions and
+                # track them down.
+                exception_count.labels('Exception').inc()
```
Comment (gfr10598) [inline exitstats.py]: Done

File: exitstats.py [hunk 3]
```diff
@@ -17,214 +19,258 @@
...
+    while True:
+        conns = agent.all_connections()
+        newclosed = []
+        for c in conns:
+            try:
+                if c.read('State') == 1:
+                    newclosed.append(c.cid)
+                    if not c.cid in closed:
+                        stats_writer.logConnection(c)
+            except Exception, e:
```
Comment (gfr10598) [inline exitstats.py]: Done

File: exitstats.py (hunk 1 above)
Comment (gfr10598) [inline exitstats.py]: Ignoring since we hope this will die soon.

File: exitstats_test.py [hunk 4]
```diff
@@ -34,150 +70,305 @@ def copy(self):
...
-        print('-', end="")
+      except urllib2.URLError:
         time.sleep(0.01)
     else:
       raise urllib2.URLError('Page not found')
 
-    self.assertTrue(
-        "sidestream_connection_count{source=\"ipv4\"} 1.0" in response, response)
+    rex = re.compile(
+        '^sidestream_connection_count[{]lsb=\"33\",type=\"ipv4\"[}] (.*)$', re.M )
```
Comment (gfr10598) [inline exitstats_test.py]: Done

File: exitstats_test.py [hunk 5]
```diff
@@ -34,150 +70,305 @@ def copy(self):
...
+      import Web100
+      print('skipping main test, because it would run forever.')
+    except ImportError:
+      # Go ahead and test main...
+      try:
+        exitstats.main(("exitstats", "server"))
+      except NameError, e:
+        pass  # This is the expected behavior
+
+    # TODO(gfr) Consider adding tests for incorrect main arguments.
```
Comment (gfr10598) [inline exitstats_test.py]: My view on this is that issues that really should be addressed should have a github issue, but that minor suggestions that someone might address but will never be high priority should just get a lightweight TODO.  Leaving as is, but will keep in mind for future TODOs.

**Pull request merged**